File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.js
function solution(gems) {
  let answer = [0, gems.length];
  const totalSize = new Set(gems).size;
    
  let left = 0, right = 0;
  const gemsCount = new Map();
  gemsCount.set(gems[0], 1);

  while (left < gems.length && right < gems.length) {
    if (gemsCount.size === totalSize) {
      // 더 짧은 경우에만 갱신(같을 때는 X)
      if (right - left < answer[1] - answer[0]) answer = [left + 1, right + 1];

      // left 늘려서 범위 좁히기
      gemsCount.set(gems[left], gemsCount.get(gems[left]) - 1);
      if (gemsCount.get(gems[left]) === 0) gemsCount.delete(gems[left]);
        
      left++;
    } else {
      // 모든 보석 포함할 때까지 right 늘려서 범위 넓히기
      right++;
      if (right >= gems.length) break;
      gemsCount.set(gems[right], (gemsCount.get(gems[right]) || 0) + 1);
    }
  }

  return answer;
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
def solution(gems):
    answer = [1, len(gems)]
    gem_kinds = len(set(gems))
    start, end = 0, 0
    gems_count = {}
    gems_count[gems[0]] = 1
    
    while start < len(gems) and end < len(gems):
        if len(gems_count) == gem_kinds:
            if end - start < answer[1] - answer[0]:
                answer = [start + 1, end + 1]
            
            gems_count[gems[start]] -= 1

            if gems_count[gems[start]] == 0:
                del gems_count[gems[start]]
            
            start += 1
        else:
            end += 1
            if end >= len(gems):
                break
            
            if gems[end] in gems_count:
                gems_count[gems[end]] += 1
            else:
                gems_count[gems[end]] = 1
    
    return answer

    